x-pack/plugins/ml/public/application/jobs/new_job/common/job_creator/geo_job_creator.ts

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { DataView } from '@kbn/data-views-plugin/public';
+import { SavedSearchSavedObject } from '../../../../../../common/types/kibana';
+import { JobCreator } from './job_creator';
+import {
+  Field,
+  Aggregation,
+  SplitField,
+  AggFieldPair,
+} from '../../../../../../common/types/fields';
+import { Job, Datafeed, Detector } from '../../../../../../common/types/anomaly_detection_jobs';
+import { createBasicDetector } from './util/default_configs';
+import { JOB_TYPE, CREATED_BY_LABEL } from '../../../../../../common/constants/new_job';
+import { getRichDetectors } from './util/general';
+import { isSparseDataJob } from './util/general';
+
+export class GeoJobCreator extends JobCreator {
+  private _geoField: Field | null = null;
+  private _geoAgg: Aggregation | null = null;
+  // set partitionField as the default split field for geo jobs
+  private _splitField: SplitField = null;
+
+  protected _type: JOB_TYPE = JOB_TYPE.GEO;
+
+  constructor(indexPattern: DataView, savedSearch: SavedSearchSavedObject | null, query: object) {
+    super(indexPattern, savedSearch, query);
+    this.createdBy = CREATED_BY_LABEL.GEO;
+    this._wizardInitialized$.next(true);
+  }
+
+  public setDefaultDetectorProperties(geo: Aggregation | null) {
+    if (geo === null) {
+      throw Error('lat_long aggregations missing');
+    }
+    this._geoAgg = geo;
+  }
+
+  public set geoField(field: Field | null) {
+    this._geoField = field;
+  }
+
+  public get geoField() {
+    return this._geoField;
+  }
+
+  public get geoAgg() {
+    return this._geoAgg;
+  }
+
+  public setGeoField(field: Field | null) {
+    this._geoField = field;
+
+    if (field === null) {
+      this.removeSplitField();
+      this._removeDetector(0);
+      this._detectors.length = 0;
+      this._fields.length = 0;
+      return;
+    }
+    const agg = this._geoAgg!;
+    if (this._detectors.length === 0) {
+      this.addDetector(agg, field);
+    } else {
+      this._editDetector(this._detectors[0], agg, field, 0);
```

Comment:
this has bug where the same detector is being re-added if the geo field selection changes.
The UI looks like it is using the new field, but he actual job JSON underneath is still using the first selection.

```suggestion
      const dtr = this._createDetector(agg, field);
      this._editDetector(dtr, agg, field, 0);
```